Source/Core/Core/PowerPC/PPCCache.cpp

```diff
@@ -94,134 +94,270 @@ InstructionCache::~InstructionCache()
+  if (addr & CACHE_VMEM_BIT)
   {
-    t = lookup_table_vmem[(addr >> 5) & 0xfffff];
+    way = lookup_table_vmem[(addr >> 5) & 0xfffff];
   }
-  else if (addr & ICACHE_EXRAM_BIT)
+  else if (addr & CACHE_EXRAM_BIT)
   {
-    t = lookup_table_ex[(addr >> 5) & 0x1fffff];
+    way = lookup_table_ex[(addr >> 5) & 0x1fffff];
   }
   else
   {
-    t = lookup_table[(addr >> 5) & 0xfffff];
+    way = lookup_table[(addr >> 5) & 0xfffff];
   }
 
-  if (t == 0xff)  // load to the cache
+  // load to the cache
+  if (!locked && way == 0xff)
   {
-    if (HID0.ILOCK)  // instruction cache is locked
-      return memory.Read_U32(addr);
+    u32 tag = addr >> 12;
+
     // select a way
     if (valid[set] != 0xff)
-      t = s_way_from_valid[valid[set]];
+      way = s_way_from_valid[valid[set]];
     else
-      t = s_way_from_plru[plru[set]];
-    // load
-    memory.CopyFromEmu(reinterpret_cast<u8*>(data[set][t].data()), (addr & ~0x1f), 32);
-    if (valid[set] & (1 << t))
+      way = s_way_from_plru[plru[set]];
+
+    if (valid[set] & (1 << way))
     {
-      if (tags[set][t] & (ICACHE_VMEM_BIT >> 12))
-        lookup_table_vmem[((tags[set][t] << 7) | set) & 0xfffff] = 0xff;
-      else if (tags[set][t] & (ICACHE_EXRAM_BIT >> 12))
-        lookup_table_ex[((tags[set][t] << 7) | set) & 0x1fffff] = 0xff;
+      // store the cache back to main memory
+      if (wrote[set] & (1 << way))
+        memory.CopyToEmu(addrs[set][way], reinterpret_cast<u8*>(data[set][way].data()), 32);
+
+      if (tags[set][way] & (CACHE_VMEM_BIT >> 12))
+        lookup_table_vmem[((tags[set][way] << 7) | set) & 0xfffff] = 0xff;
+      else if (tags[set][way] & (CACHE_EXRAM_BIT >> 12))
+        lookup_table_ex[((tags[set][way] << 7) | set) & 0x1fffff] = 0xff;
       else
-        lookup_table[((tags[set][t] << 7) | set) & 0xfffff] = 0xff;
+        lookup_table[((tags[set][way] << 7) | set) & 0xfffff] = 0xff;
     }
 
-    if (addr & ICACHE_VMEM_BIT)
-      lookup_table_vmem[(addr >> 5) & 0xfffff] = t;
-    else if (addr & ICACHE_EXRAM_BIT)
-      lookup_table_ex[(addr >> 5) & 0x1fffff] = t;
+    // load
+    memory.CopyFromEmu(reinterpret_cast<u8*>(data[set][way].data()), (addr & ~0x1f), 32);
+
+    if (addr & CACHE_VMEM_BIT)
+      lookup_table_vmem[(addr >> 5) & 0xfffff] = way;
+    else if (addr & CACHE_EXRAM_BIT)
+      lookup_table_ex[(addr >> 5) & 0x1fffff] = way;
     else
-      lookup_table[(addr >> 5) & 0xfffff] = t;
-    tags[set][t] = tag;
-    valid[set] |= (1 << t);
+      lookup_table[(addr >> 5) & 0xfffff] = way;
+    tags[set][way] = tag;
+    addrs[set][way] = addr;
+    valid[set] |= (1 << way);
+    wrote[set] &= ~(1 << way);
   }
+
   // update plru
-  plru[set] = (plru[set] & ~s_plru_mask[t]) | s_plru_value[t];
-  const u32 res = Common::swap32(data[set][t][(addr >> 2) & 7]);
-  const u32 inmem = memory.Read_U32(addr);
-  if (res != inmem)
+  if (way != 0xff)
+    plru[set] = (plru[set] & ~s_plru_mask[way]) | s_plru_value[way];
```

Comment:
It seems a bit weird that this happens even if `locked` is set (including for invalidate). I'm not sure if this matters in practice though, and the manual doesn't specify one way or the other.

Reply:
I don't know how real hardware acts here for sure (and can't think of a good way to test it), but that wasn't intentional and I fixed it now